src/spsc_queue.ml

```diff
@@ -36,29 +37,61 @@ let create ~size_exponent =
     head = Atomic.make 0;
     tail = Atomic.make 0;
     mask = size - 1;
-    array = Array.init size (fun _ -> None);
+    array = Array.make size None;
+    lock = Atomic.make None;
   }
 
-let push { array; head; tail; mask; _ } element =
+let unlock lock =
+  if Atomic.get lock <> None then
+    match Atomic.exchange lock None with
+    | Some release -> release ()
+    | None -> ()
+
+let try_push { array; head; tail; mask; lock } element =
   let size = mask + 1 in
-  let head_val = Atomic.get head in
   let tail_val = Atomic.get tail in
-  if head_val + size == tail_val then raise Full
+  let head_val = Atomic.get head in
+  if head_val + size = tail_val then false
   else (
     Array.set array (tail_val land mask) (Some element);
-    Atomic.set tail (tail_val + 1))
+    Atomic.set tail (tail_val + 1);
+    unlock lock;
+    true)
 
-let pop { array; head; tail; mask; _ } =
+let try_pop { array; head; tail; mask; lock } =
   let head_val = Atomic.get head in
   let tail_val = Atomic.get tail in
-  if head_val == tail_val then None
+  if head_val = tail_val then None
   else
     let index = head_val land mask in
     let v = Array.get array index in
     (* allow gc to collect it *)
     Array.set array index None;
     Atomic.set head (head_val + 1);
     assert (Option.is_some v);
+    unlock lock;
     v
 
 let size { head; tail; _ } = Atomic.get tail - Atomic.get head
+
+let wait ~rdv t expected_size =
+  let release, wait = rdv () in
+  let some_release = Some release in
+  if Atomic.compare_and_set t.lock None some_release then
+    if size t = expected_size then wait ()
+    else
+      let (_ : bool) = Atomic.compare_and_set t.lock some_release None in
+      ()
+  else unlock t.lock
```

Comment:
Hmm... Did I understand correctly that if two calls to `wait` are being made, e.g. to `pop`s when the queue is empty, then what happens is that the second call to `wait` actually wakes up the first one?  So, in that case you get a very expensive busy wait?

Reply:
Ah.. SPSC! Right.

Reply:
Yeah :P This was a quick example that only works for SPSC, but in general multiple waiting `pop` would actually push their "release" function onto a queue. Then the next `push` would see the value queue as empty and pop/wake-up the first available `pop`.